CDAP-9148 refactoring to wrap all pipeline plugins

albertshau opened the pull request:
Fixes CDAP-9418 and HYDRATOR-122.
Refactoring so that any known plugin type instantiated by the
pipeline is wrapped so that the context classloader is set
before any plugin method is called, and so that all certain
methods are timed and a metrics emitted for time spent.
Also performs one part of the fix to stage logging, which is to
set a key in the logging MDC before any plugin method is called.

albertshau committed:
CDAP-9148 refactoring to wrap all pipeline plugins

Refactoring so that any known plugin type instantiated by the
pipeline is wrapped so that the context classloader is set
before any plugin method is called, and so that all certain
methods are timed and a metrics emitted for time spent.
Also performs one part of the fix to stage logging, which is to
set a key in the logging MDC before any plugin method is called.

albertshau commented:
http://builds.cask.co/browse/CDAP-DUT5619

albertshau committed:
CDAP-9148 add flags to disable process timing

Adding a flag to disable process timing. Also includes some
refactoring to make sure the flag is available in all the places
that instantiate plugins.

yaojiefeng commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultStageMetrics.java:
```diff
@@ -34,20 +35,22 @@
   // Only used by Externalizable
   public DefaultStageMetrics() {
-    // no-op
+    this(null, "");
```
Is there any reason that we need this?

albertshau replied:
it's called by spark to deserialize the class.

yaojiefeng commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/plugin/TimingCaller.java:
```diff
@@ -0,0 +1,91 @@
+    double mean = (double) sumOfValues / numValues;
+    double squareOfMean = mean * mean;
+    double meanOfSquares = (double) sumOfSquaresOfValues / numValues;
+    double stddev = Math.sqrt(meanOfSquares - squareOfMean);
```
The standard deviation here seems wrong. According to https://en.wikipedia.org/wiki/Standard_deviation, what we need is the sum of (actual - mean)^2

albertshau replied:
We can't store all the different data points, and this is a mathematically equivalent calculation. See https://math.stackexchange.com/a/102982 for an explanation.

albertshau replied:
though i originally was storing millis instead of micros. I think there will be an overflow concern with this, so i'll switch it to https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm

yaojiefeng commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java:
```diff
@@ -1,3 +1,4 @@
+
```
remove this line

yaojiefeng commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/plugin/SparkPipelinePluginContext.java:
```diff
@@ -0,0 +1,54 @@
+  protected Object wrapUnknownPlugin(String pluginId, Object plugin, Caller caller) {
+    if (plugin instanceof Windower) {
+      return new WrappedWindower((Windower) plugin, caller);
+    } else if (plugin instanceof SparkCompute) {
```
I am not sure I understand this correctly. But SparkCompute and SparkSink are valid plugin types for data pipeline app, which means they can be run under mapreduce engine. Will these plugins be wrapped correctly for mapreduce?

albertshau replied:
these plugins cannot be run by mapreduce. If a pipeline uses them, everything is converted into a single spark program.

yaojiefeng replied:
I see.

yaojiefeng commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/plugin/WrappedSparkCompute.java:
```diff
@@ -0,0 +1,74 @@
+
+  @Override
+  public JavaRDD<OUT> transform(final SparkExecutionPluginContext context, final JavaRDD<IN> input) throws Exception {
+    return caller.call(new Callable<JavaRDD<OUT>>() {
```
Do we need `CallArgs.TRACK_TIME` to track the time of this method?

albertshau replied:
it is intentionally not tracked because this method only runs once in the spark driver. It is basically spark's equivalent of building an execution plan.

yaojiefeng commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/plugin/WrappedSparkSink.java:
```diff
@@ -0,0 +1,87 @@
+
+  @Override
+  public void run(final SparkExecutionPluginContext context, final JavaRDD<IN> input) throws Exception {
+   caller.call(new Callable<Void>() {
```
Do we need CallArgs.TRACK_TIME to track the time of this method?

albertshau replied:
similar reason here, it would be one very short method call from the spark driver, and not the actual amount of time spent by the spark executors.

yaojiefeng commented on cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/plugin/WrappedStreamingSource.java:
```diff
@@ -0,0 +1,63 @@
+
+  @Override
+  public JavaDStream<T> getStream(final StreamingContext context) throws Exception {
+    return caller.call(new Callable<JavaDStream<T>>() {
```
Do we need CallArgs.TRACK_TIME to track the time of this method?

albertshau replied:
similar reason here, except this only gets called once the very first time the streaming pipeline runs, and never again after that.

yaojiefeng commented:
Also, just curious, for the newly added sparkprogram plugin, do we want to add time metrics to that?

yaojiefeng commented:
Just a few comments and a question.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/plugin/TimingCaller.java:
```diff
@@ -0,0 +1,91 @@
+    double mean = (double) sumOfValues / numValues;
+    double squareOfMean = mean * mean;
+    double meanOfSquares = (double) sumOfSquaresOfValues / numValues;
+    double stddev = Math.sqrt(meanOfSquares - squareOfMean);
```
ironically that wiki page had a bug in its algorithm, which i just fixed

albertshau commented:
similar thing for the sparkprogram plugin, everything we could time would be in the driver and not in the executors

albertshau committed:
CDAP-9148 wrap sparkprogram plugins and change stddev calculation

Change the standard deviation calculation to avoid overflow
concerns. Also wrapping the new sparkprogram plugins so that they
at least get the logging wrapping.

yaojiefeng commented:
LGTM
